### DISCUSSION TITLE
How to (imperatively) map a list of simple data types to the DB?

### DISCUSSION THREAD

**Discussion by petarlishov** [sqlalchemy/sqlalchemy#11865] [Usage Questions] [answered]
I am trying to set up the ORM mapping of a Python class structure that looks like this:

```python
class SomeClass:
   id: int
   elements: list[<simple data type like a string, integer or ideally an enum if possible>]

class ExampleEnum: # not necessary that we rely on enums
   A = "A"
   B = "B"
```

I am struggling to find the correct way to map the above onto database tables using SQLAlchemy 2.0 (how to define mappings imperatively in particular)

I am not sure if this will consist of two tables:
```sql
create table some_class  (
  `id` int NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
)

create table some_class_element  (
  `some_class_id` int NOT NULL,
  `element` int NOT NULL,
  PRIMARY KEY (`some_class_id`)
)
```

or three tables:
```sql
create table some_class  (
  `id` int NOT NULL AUTO_INCREMENT,
  PRIMARY KEY (`id`)
)

create table element  (
  `id` int NOT NULL AUTO_INCREMENT,
  `value` int NOT NULL,
  PRIMARY KEY (`id`)
)

create table some_class_element  (
  `some_class_id` int NOT NULL,
  `some_element_id` int NOT NULL,
  PRIMARY KEY (`some_class_id`, `some_element_id`)
)
```

I feel like the two-table approach makes more sense simply from a database structure and avoids unnecessary joins, but have seen some comments stating this might need three for the purposes of ORM mapping with SQLAlchemy.

Basically, am I meant to model the DB with two or three tables? And how would the imperative mapping code look in that case?

**Comment by zzzeek** [comment]
two or three tables here depends on how you want to represent the enum.   if you want each enum value to be stored in a dedicated lookup table, that's the three table version.  if OTOH you want to indicate the enum as a real ENUM SQL datatype, or a CHAR or integer that's directly stored as the record, then you'd use two tables.

to get the list of elements as a single enum/string/int you would use the [association proxy](https://docs.sqlalchemy.org/en/20/orm/extensions/associationproxy.html#simplifying-scalar-collections).

the first Q to be answered is how do you want the ENUM represented in the database.  if using MySQL, MariaDB or PostgreSQL you have the option to use a native ENUM datatype.

**Reply by CaselIT** [reply]
In postgresql you could also use a one table approach using array (or even json)

**Reply by petarlishov** [reply, marked as answer]
OK I have been attempting the association proxy approach, and I finally got to a working solution (simplified from the original so I might be messing up some of these in rewriting them):

First package with the classes I am looking to map
```python
class ExampleEnum: # not necessary that we rely on enums
    A = "A"
    B = "B"

class SomeClass:
    id: int
    elements: set[ExampleEnum]

    def __init__(elements: set[ExampleEnum]):
        if not elements:
            elements = set()
        self.elements = elements

```

Second package with the mapping attempt:
```python
class SomeClassToExampleEnum:
    some_id: int
    example: ExampleEnum

    def __init__(self, example: ExampleEnum):
        self.example = example


def _map(self, mapper: registry, schema: str):
    some_class_table = Table(
        "some_class",
        mapper.metadata,
        Column("id", Integer, nullable=False, primary_key=True),
        schema=schema,
    )

    some_class_to_example_enum_table = Table(
        "some_class_to_example_enum",
        mapper.metadata,
        Column("some_id", ForeignKey(some_class_table.c.id), primary_key=True),
        Column("example", Enum(ExampleEnum, native_enum=False, length=100), primary_key=True),
        schema=schema,
    )

    mapper.map_imperatively(SomeClassToExampleEnum, some_class_to_example_enum_table)
    mapper.map_imperatively(
        SomeClass,
        some_class_table,
        properties={
            "_elements": relationship(
                SomeClassToExampleEnum,
                collection_class=set,
                cascade="all, delete-orphan",
            ),
        },
    )
    SomeClass.elements = association_proxy("_elements", "elements")
```

Example test:
```python
def test_persistence(engine):
    some_id: int

    test_object = SomeClass(elements={ExampleEnum.A, ExampleEnum.B})
 
    with Session(bind=engine) as session, session.begin():
        session.add(test_object)
        session.flush()
        some_id = test_object.id

    with Session(bind=engine) as session, session.begin():
        fetched = session.get(SomeClass, some_id)
```
